File: frontend/src/api/delivery.ts
import { apiClient } from './client'

export type BatchStatus = 'pending' | 'dispatched' | 'completed' | 'cancelled'
export type DeliveryStatus = 'pending' | 'in_transit' | 'delivered' | 'failed'

export interface DeliveryOrderItem {
  id: number
  order_id: number
  batch_id?: number
  delivery_address: string
  status: DeliveryStatus
  delivery_notes?: string
  delivered_at?: string
}

export interface DeliveryBatch {
  id: number
  batch_number: string
  driver_name?: string
  driver_phone?: string
  status: BatchStatus
  notes?: string
  scheduled_at?: string
  dispatched_at?: string
  completed_at?: string
  created_at: string
  delivery_orders: DeliveryOrderItem[]
}

export interface DeliveryBatchCreate {
  driver_name?: string
  driver_phone?: string
  notes?: string
  scheduled_at?: string
  order_ids?: number[]
}

export const deliveryApi = {
  listBatches: (skip = 0, limit = 50) =>
    apiClient.get<DeliveryBatch[]>('/delivery/batches', { params: { skip, limit } }).then((r) => r.data),

  createBatch: (data: DeliveryBatchCreate) =>
    apiClient.post<DeliveryBatch>('/delivery/batches', data).then((r) => r.data),

  getBatch: (id: number) =>
    apiClient.get<DeliveryBatch>(`/delivery/batches/${id}`).then((r) => r.data),

  updateBatch: (id: number, data: Partial<DeliveryBatchCreate & { status: BatchStatus }>) =>
    apiClient.patch<DeliveryBatch>(`/delivery/batches/${id}`, data).then((r) => r.data),

  assignOrders: (batchId: number, order_ids: number[]) =>
    apiClient.post<DeliveryBatch>(`/delivery/batches/${batchId}/assign`, { order_ids }).then((r) => r.data),

  suggestBatches: (max_per_batch = 10) =>
    apiClient.get<number[][]>('/delivery/batches/suggest', { params: { max_per_batch } }).then((r) => r.data),

  getUnassigned: () =>
    apiClient.get<DeliveryOrderItem[]>('/delivery/orders/unassigned').then((r) => r.data),
}

File: backend/app/crud/delivery.py
from datetime import datetime, timezone
from sqlalchemy import select
from sqlalchemy.orm import selectinload
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.delivery import DeliveryBatch, DeliveryOrder, BatchStatus
from app.schemas.delivery import DeliveryBatchCreate, DeliveryBatchUpdate, AssignOrdersToBatch


async def _generate_batch_number(db: AsyncSession) -> str:
    result = await db.execute(select(DeliveryBatch).order_by(DeliveryBatch.id.desc()).limit(1))
    last = result.scalar_one_or_none()
    next_id = (last.id + 1) if last else 1
    return f"BATCH-{datetime.now(timezone.utc).strftime('%Y%m%d')}-{next_id:04d}"


async def get_batches(db: AsyncSession, skip: int = 0, limit: int = 50) -> list[DeliveryBatch]:
    result = await db.execute(
        select(DeliveryBatch)
        .options(selectinload(DeliveryBatch.delivery_orders))
        .order_by(DeliveryBatch.created_at.desc())
        .offset(skip)
        .limit(limit)
    )
    return list(result.scalars().all())


async def get_batch(db: AsyncSession, batch_id: int) -> DeliveryBatch | None:
    result = await db.execute(
        select(DeliveryBatch)
        .options(selectinload(DeliveryBatch.delivery_orders))
        .where(DeliveryBatch.id == batch_id)
    )
    return result.scalar_one_or_none()


async def create_batch(db: AsyncSession, data: DeliveryBatchCreate) -> DeliveryBatch:
    batch_number = await _generate_batch_number(db)
    batch = DeliveryBatch(
        batch_number=batch_number,
        driver_name=data.driver_name,
        driver_phone=data.driver_phone,
        notes=data.notes,
        scheduled_at=data.scheduled_at,
    )
    db.add(batch)
    await db.flush()

    if data.order_ids:
        await _assign_orders(db, batch.id, data.order_ids)

    await db.commit()
    await db.refresh(batch)
    return batch


async def update_batch(db: AsyncSession, batch: DeliveryBatch, data: DeliveryBatchUpdate) -> DeliveryBatch:
    for field, value in data.model_dump(exclude_unset=True).items():
        setattr(batch, field, value)

    if data.status == BatchStatus.dispatched and not batch.dispatched_at:
        batch.dispatched_at = datetime.now(timezone.utc)
    if data.status == BatchStatus.completed and not batch.completed_at:
        batch.completed_at = datetime.now(timezone.utc)

    await db.commit()
    await db.refresh(batch)
    return batch


async def _assign_orders(db: AsyncSession, batch_id: int, order_ids: list[int]) -> None:
    result = await db.execute(
        select(DeliveryOrder).where(DeliveryOrder.order_id.in_(order_ids))
    )
    delivery_orders = result.scalars().all()
    for do in delivery_orders:
        do.batch_id = batch_id


async def assign_orders_to_batch(
    db: AsyncSession, batch: DeliveryBatch, data: AssignOrdersToBatch
) -> DeliveryBatch:
    await _assign_orders(db, batch.id, data.order_ids)
    await db.commit()
    await db.refresh(batch)
    return batch


async def get_unassigned_delivery_orders(db: AsyncSession) -> list[DeliveryOrder]:
    result = await db.execute(
        select(DeliveryOrder).where(DeliveryOrder.batch_id.is_(None))
    )
    return list(result.scalars().all())
